Climber

JoeJanickiChSh opened the pull request:
Added climber subsystem and telescope map which extends and retracts two telescoping arms and stops when a limit switch is hit.

JoeJanickiChSh committed:
Started to implement climber functionality

Started to implement climber functionality

JoeJanickiChSh committed:
Added ArmMap and ArmGroup to RobotMap

Added a new map for an arm and a group of arms for the climber.

JoeJanickiChSh committed:
Basic button bindings

JoeJanickiChSh committed:
Move functionality out of RobotMap

JoeJanickiChSh committed:
Change commands for extending/retracting

JoeJanickiChSh committed:
Add Gaston Map

JoeJanickiChSh committed:
Change whileHeld to whenPressed

JoeJanickiChSh committed:
Add commands for ignoring limit switches

Commands which extend/retract without checking limit switches in case switches stop working.

JoeJanickiChSh committed:
Fix button bindings for ignoring limits

JoeJanickiChSh committed:
Add some comments

JoeJanickiChSh committed:
Change whenPressed to whileHeld for button bindings

This should work for startEnd commands now

msoucy commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -1,17 +1,45 @@
 
 public class Robot extends CommandRobot {
 
+  private Command autonomousCommand;
```
This and the chooser are not necessary - for `CommandRobot`, the only thing you need to do in order to make an autonomous is:

```java
@Autonomous
public final CommandBase auto1 = foobar();
```

msoucy commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -1,17 +1,45 @@
+
+  private final RobotMap map = getMapForName("Gaston", RobotMap.class, "frc.robot.maps");
+
+  private final Climber leftClimber = new Climber(map.getLeftTelescopeMap());
```
Is there any situation when you would want the climbers to be doing different things?

AidenColhon replied:
Each arm has seperate motors and limit switches. When one reaches it's limit it is good to act on it's own.

msoucy replied:
👍

msoucy commented on src/main/java/frc/robot/subsystems/Climber.java:
```diff
@@ -0,0 +1,77 @@
+
+  // Constants:
+  private final double EXTEND_SPEED = 1.0;
+  private final double RETRACT_SPEED = 1.0;
```
It might be helpful to define a single direction as "forward" and make both `EXTEND_SPEED` and `RETRACT_SPEED` use that same direction for positive.

msoucy replied:
I meant set `EXTEND_SPEED = 1.0` and `RETRACT_SPEED = -1.0`, not to introduce a new `UP`.

msoucy commented on src/main/java/frc/robot/subsystems/Climber.java:
```diff
@@ -0,0 +1,77 @@
+
+  @Override
+  public void periodic() {
+    // This method will be called once per scheduler run
```
Function is not necessary, you can just remove this if it has no contents

msoucy commented on src/main/java/frc/robot/subsystems/Climber.java:
```diff
@@ -0,0 +1,77 @@
+import com.chopshop166.chopshoplib.commands.SmartSubsystemBase;
+import com.chopshop166.chopshoplib.motors.SmartMotorController;
+
+import java.util.function.BooleanSupplier;
```
Please use alt-shift-O to organize the imports

msoucy commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -1,17 +1,45 @@
   @Override
   public void robotInit() {
-    // Nothing to initialize
+
```
This is definitely a mistake on my part, but please have `robotInit` call `super.robotInit()` inside it. Otherwise nothing will run.

bot190 commented on src/main/java/frc/robot/subsystems/Climber.java:
```diff
@@ -0,0 +1,77 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
```
Please remove the WPILib copyright notice.

bot190 commented on src/main/java/frc/robot/maps/GastonMap.java:
```diff
@@ -0,0 +1,28 @@
+    private final DigitalInput rightLowerLimit = new DigitalInput(3);
+
+    public TelescopeMap getLeftTelescopeMap() {
+        TelescopeMap map = new TelescopeMap(new PIDSparkMax(0, MotorType.kBrushless), leftUpperLimit::get,
```
Let's create the `PIDSparkMax` object outside of this like you did for the Digital Inputs. We're likely to need to adjust settings on them.

JoeJanickiChSh committed:
Remove unnecessary code and organize imports

Remove unnecessary code, organize imports, add super.robotInit(), and add UP constant for the climber subsystem.

JoeJanickiChSh committed:
Move motor defenitions out of methods on GastonMap

bot190 commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -1,17 +1,39 @@
+    controller.b().whileHeld(parallel("Retract", leftClimber.retract(), rightClimber.retract()));
+
+    // Button bindings for ignoring limit switches
+    controller.x()
```
These should probably be on the d-pad where they're harder to press accidentally

bot190 commented on src/main/java/frc/robot/subsystems/Climber.java:
```diff
@@ -0,0 +1,69 @@
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.maps.RobotMap.TelescopeMap;
+
+public class Climber extends SmartSubsystemBase {
```
We should probably have a command that controls the elevator with the triggers so we can try different speeds easily.

msoucy replied:
This exists now so I'll resolve this

bot190 commented on src/main/java/frc/robot/subsystems/Climber.java:
```diff
@@ -0,0 +1,69 @@
+    lowerLimit = map.getLowerLimit();
+  }
+
+  public CommandBase extend() {
```
These should also use the `motors.Modifier.upperLimit` and `motors.Modifier.lowerLimit` to ensure that the motors don't try to drive against the limit switches. This should solve the issue we talked about where the command will run briefly before checking `isFinished`.

bot190 replied:
See the code here https://github.com/chopshop-166/chopshoplib/blob/main/core/src/main/java/com/chopshop166/chopshoplib/motors/Modifier.java

msoucy replied:
You could probably move the modifiers and group into the class itself so you're not creating a new one every time

JoeJanickiChSh committed:
Add variable speed for controlling telescopes

msoucy commented on src/main/java/frc/robot/maps/GastonMap.java:
```diff
@@ -0,0 +1,29 @@
+
+    public TelescopeMap getLeftTelescopeMap() {
+        TelescopeMap map = new TelescopeMap(leftMotor, leftUpperLimit::get, leftLowerLimit::get);
+        return map;
```
Useless temporary variable

JoeJanickiChSh committed:
Move modifiers out of commands

JoeJanickiChSh committed:
Use updated ModifierGroup constructor
